### PR TITLE
Basic support macro 2.0

### PR DESCRIPTION
**Pull request by edwin0cheng**
Turn out it is quite straight forward :)

r @jonas-schievink 

m2

### CODE REVIEW

File: crates/hir_def/src/nameres/tests/macros.rs [hunk 1]
```diff
@@ -837,3 +837,25 @@ fn collects_derive_helpers() {
...
+//- /nested.rs
+structs!(Bar, Baz);
+"#,
+        expect![[r#"
+            crate
+            Foo: t
+            structs: m
+        "#]],
+    );
+}
```
Comment (bjorn3) [inline crates/hir_def/src/nameres/tests/macros.rs:861]: Can you also add a test that something like the following doesn't resolve the item outside the macro?

```rust
//- /lib.rs
pub macro struct() {
    struct Foo { field: u32 }
}
struct!();
```

Comment (edwin0cheng) [reply crates/hir_def/src/nameres/tests/macros.rs:861]: We still don't have macro hygiene yet, so `Foo` will be valid in your case anyway...

**Review by jonas-schievink — COMMENTED** [review]
Very nice!

File: crates/hir_def/src/nameres/tests/macros.rs [hunk 2]
```diff
@@ -837,3 +837,25 @@ fn collects_derive_helpers() {
...
+        r#"
+//- /lib.rs
+pub macro structs($($i:ident),*) {
+    $(struct $i { field: u32 } )*
+}
+
+structs!(Foo);
+
+//- /nested.rs
+structs!(Bar, Baz);
```
[the remark this answers is not in the document]
Comment (edwin0cheng) [reply crates/hir_def/src/nameres/tests/macros.rs]: Removed.

**Pull request merged**